web_src/js/components/RepoBranchTagSelector.vue

```diff
@@ -0,0 +1,293 @@
+        this.isViewTree = false;
+        this.isViewTag = false;
+        this.isViewBranch = false;
+        this.$refs.dropdownRefName.textContent = item.name;
+        if (this.setAction) {
+          $(`#${this.branchForm}`).attr('action', url);
+        } else {
+          $(`#${this.branchForm} input[name="refURL"]`).val(url);
+        }
+        $(`#${this.branchForm} input[name="ref"]`).val(item.name);
+        if (item.tag) {
+          this.isViewTag = true;
+          $(`#${this.branchForm} input[name="refType"]`).val('tag');
+        } else {
+          this.isViewBranch = true;
+          $(`#${this.branchForm} input[name="refType"]`).val('branch');
+        }
+        if (this.submitForm) {
+          $(`#${this.branchForm}`).trigger('submit');
+        }
+        this.menuVisible = false;
+      }
+    },
+    createNewBranch() {
+      if (!this.showCreateNewBranch) return;
+      $(this.$refs.newBranchForm).trigger('submit');
+    },
+    focusSearchField() {
+      nextTick(() => {
+        this.$refs.searchField.focus();
+      });
+    },
+    getSelected() {
+      for (let i = 0, j = this.items.length; i < j; ++i) {
+        if (this.items[i].selected) return this.items[i];
+      }
+      return null;
+    },
+    getSelectedIndexInFiltered() {
+      for (let i = 0, j = this.filteredItems.length; i < j; ++i) {
+        if (this.filteredItems[i].selected) return i;
+      }
+      return -1;
+    },
+    scrollToActive() {
+      let el = this.$refs[`listItem${this.active}`];
+      if (!el || !el.length) return;
+      if (Array.isArray(el)) {
+        el = el[0];
+      }
+
+      const cont = this.$refs.scrollContainer;
+      if (el.offsetTop < cont.scrollTop) {
+        cont.scrollTop = el.offsetTop;
+      } else if (el.offsetTop + el.clientHeight > cont.scrollTop + cont.clientHeight) {
+        cont.scrollTop = el.offsetTop + el.clientHeight - cont.clientHeight;
+      }
+    },
+    keydown(event) {
+      if (event.keyCode === 40) { // arrow down
+        event.preventDefault();
+
+        if (this.active === -1) {
+          this.active = this.getSelectedIndexInFiltered();
+        }
+
+        if (this.active + (this.showCreateNewBranch ? 0 : 1) >= this.filteredItems.length) {
+          return;
+        }
+        this.active++;
+        this.scrollToActive();
+      } else if (event.keyCode === 38) { // arrow up
+        event.preventDefault();
+
+        if (this.active === -1) {
+          this.active = this.getSelectedIndexInFiltered();
+        }
+
+        if (this.active <= 0) {
+          return;
+        }
+        this.active--;
+        this.scrollToActive();
+      } else if (event.keyCode === 13) { // enter
+        event.preventDefault();
+
+        if (this.active >= this.filteredItems.length) {
+          this.createNewBranch();
+        } else if (this.active >= 0) {
+          this.selectItem(this.filteredItems[this.active]);
+        }
+      } else if (event.keyCode === 27) { // escape
```

Comment:
```suggestion
      } else if (event.code === 'Escape') {
```

Reply:
This PR only moves the code into Vue SFC, **doesn't touch the unrelated logic**